DAQ/src/ArtBinaryPacketsFromDigis_module.cc

```diff
@@ -0,0 +1,1113 @@
+	// if (_diagLevel > 1) {
+	//   std::cout << "[ArtBinaryPacketsFromDigis::processTrackerData ] dtcId ="<< (int)dtcID << " rcoID = "<< (int)rocID << std::endl;
+	//   std::cout << " DTC    ROC" <<std::endl;
+	// }
+
+	for (size_t curHitIdx = 0; curHitIdx < tmpTrackerData.size(); curHitIdx++) {
+	  // if (_diagLevel > 1) {
+	  //   std::cout << (int)tmpTrackerData[curHitIdx].first.DTCID << "    "<< (int)tmpTrackerData[curHitIdx].first.ROCID <<std::endl;
+	  // }
+
+	  if (tmpTrackerData[curHitIdx].first.DTCID == dtcID &&
+	      tmpTrackerData[curHitIdx].first.ROCID == rocID) {
+	    // if (_diagLevel > 1) {
+	    //   std::cout <<"\t data found!" << std::endl;
+	    // }
+	    trackerData.push_back(tmpTrackerData[curHitIdx]);
+	    data_found = true;
+	  }
+	}
+
+	if (!data_found) {
+	  // No hits, so just fill a header packet and no data packets
+	  DataBlockHeader   headerData;
+	  TrackerDataPacket trkData;
+
+	  fillEmptyHeaderDataPacket(headerData, eventNum, rocID, dtcID, DTCLib::DTC_Subsystem_Tracker);
+	  fillEmptyTrackerDataPacket(trkData);
+
+	  trackerData.push_back(std::pair<DataBlockHeader, TrackerDataPacket>(headerData, trkData));
+	}
+      } //Done looping over the ROCs in a given DTC
+    }
+  }
+
+
+
+  flag_mask_type ArtBinaryPacketsFromDigis::filter(TrackerDataPacket& TrkData,
+						   calib_constant_type clockstart,
+						   calib_constant_type panelTDCoffset, calib_constant_type hvoffset, calib_constant_type caloffset,
+						   calib_constant_type energy_max_LSHIFT8, calib_constant_type energy_min_LSHIFT8,
+						   calib_constant_type gain_RSHIFT15,
+						   calib_constant_type inverse_ionization_energy_LSHIFT26) {
+
+    int failed_time = 0;
+    int failed_energy = 0;
+    int thistdc = TrkData.TDC0 < TrkData.TDC1 ? TrkData.TDC0 + hvoffset : TrkData.TDC1 + caloffset;
+    thistdc = thistdc + clockstart + panelTDCoffset;
+    if (thistdc < LOWER_TDC || thistdc > UPPER_TDC) {
+      failed_time = 1;
+    }
+    std::array<adc_t, 15> const waveform = TrkData.Waveform();
+    //sum up presamples
+    int pedsum = 0;
+    for (int i = 0; i < NUM_PRESAMPLES; i++) {
+      pedsum += waveform[i];
+    }
+    int pedestal = pedsum / NUM_PRESAMPLES;
+
+    int peak = 0;
+    for (int i = START_SAMPLES; i < NUM_SAMPLES; i++) {
+      if (TrkData.Waveform().at(i) > peak) {
+	peak = waveform[i];
+      }
+      else {
+	break;
+      }
+    }
+
+    int energy = peak - pedestal;
+
+    int energy_max_adjusted = ((((energy_max_LSHIFT8 * gain_RSHIFT15) >> 9)* inverse_ionization_energy_LSHIFT26) >> 10);
+    int energy_min_adjusted = ((((energy_min_LSHIFT8 * gain_RSHIFT15) >> 9)* inverse_ionization_energy_LSHIFT26) >> 10);
+    if (energy > energy_max_adjusted || energy < energy_min_adjusted) {
+      failed_energy = 1;
+    }
+
+    return (failed_energy << 1) | failed_time;
+
+  }
+
+}
+
+
+using mu2e::ArtBinaryPacketsFromDigis;
+DEFINE_ART_MODULE(ArtBinaryPacketsFromDigis);
```

Comment:
You can write this in one line:
DEFINE_ART_MODULE(mu2e::ArtBinaryPacketsFromDigis);

In the very old days it had to be done on 2 lines because of limitations in the CPP.